### PR TITLE
Add async notifications for expiring group members.

### PR DESCRIPTION
**Pull request by tmildorf**

### CODE REVIEW

**Comment by tmildorf** [comment]
Ping.

File: grouper/models.py [hunk 1]
```diff
@@ -791,7 +796,20 @@ def my_log_entries(self):
...
         return AuditLog.get_entries(self.session, on_group_id=self.id, limit=20)
 
+    def my_owners_as_strings(self):
+        """Returns a list of usernames."""
+        return self.my_owners().keys()
+
+    def my_owners(self):
+        """Returns a dictionary from username to records."""
+        od = OrderedDict()
+        for member_type_and_name, member in self.my_members().iteritems():
```
Comment (mattspitz) [inline grouper/models.py]: nit: you can further unpack this if you'd like. Not a big deal, just FYI:
`for (member_type, name), member in self.my_members().iteritems()`

File: grouper/models.py [hunk 2]
```diff
@@ -1274,6 +1367,92 @@ class AsyncNotification(Model):
...
+
+        This is the main email sending method to send out a templated email. This is used to
+        asynchronously queue up the email for sending.
+
+        Args:
+        recipients (str or list(str)): Email addresses that will receive this mail. This
+            argument is either a string (which might include comma separated email addresses)
+            or it's a list of strings (email addresses).
+        subject (str): Subject of the email.
+        body (str): Text of bodyl
```
Comment (mattspitz) [inline grouper/models.py]: nit: typo

File: grouper/models.py [hunk 3]
```diff
@@ -1274,6 +1367,92 @@ class AsyncNotification(Model):
...
+        member_name, email address) tuples.
+        """
+        tuples = []
+        emails = session.query(AsyncNotification).filter(
+            AsyncNotification.key.like("EXPIRATION%"),
+            AsyncNotification.sent == False,
+            AsyncNotification.send_after < now_ts,
+        ).all()
+        for email in emails:
+            _, group_name, member_name = email.key.split(ILLEGAL_NAME_CHARACTER)
```
Comment (mattspitz) [inline grouper/models.py]: nit: Might be best to write this as a function and put it next to where you generate the `_expiration_key` for clarity.

**Comment by mattspitz** [comment]
Some small comments, but this LGTM. Sorry it took so long!

File: grouper/models.py (hunk 3 above)
Comment (tmildorf) [inline grouper/models.py]: Agreed.

**Pull request merged**